### PR TITLE
refactor!: Change the way (C)KF can be extended

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR changes the way the extension mechanism for the (C)KF work. It switches from templating them on all of these extension structs to using the `Delegate` type to define the interface.

At the same time this PR restructures the measurement selection and internal implementation of the `filter` method. Selection should now be more efficient  (in place as much as possible) and not make use of our variant `Measurement` type anymore, but operate directly on the track states.

This PR is dependent on

- #1059
- #1060 
- #1061
and includes their changes, so this PR's changeset will reduce once they are merged.

I'll have to do some manual rebasing to undo changes needed to make the other 3 PRs standalone, so I'll keep this WIP until that's completed.

### CODE REVIEW

**Comment by paulgessinger** [comment]
@XiaocongAi all the prerequisites are now merged, so ready for review.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 1]
```diff
@@ -529,105 +571,178 @@ class CombinatorialKalmanFilter {
...
+
         // Calibrate all the source links on the surface since the selection has
         // to be done based on calibrated measurement
-        std::vector<BoundVariantMeasurement> measurements;
-        measurements.reserve(nSourcelinks);
         for (auto it = lower_it; it != upper_it; ++it) {
-          measurements.emplace_back(
-              m_calibrator(m_sourcelinkAccessor.at(it), boundParams));
+          // get the source link
+          const auto& sourceLink = m_sourcelinkAccessor.at(it);
```
Comment (XiaocongAi) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:602]: As this method is getting large, do you think we could put the creation of the track state candidates into a separate method?

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 2]
```diff
@@ -529,105 +571,178 @@ class CombinatorialKalmanFilter {
...
+
+          // now calibrate the track state
+          m_extensions.calibrator(state.geoContext, ts);
+
+          result.trackStateCandidates.push_back(ts);
         }
 
         // Invoke the measurement selector to select compatible measurements
         // with the predicted track parameter. It could return either the
         // compatible measurement indices or an outlier index.
```
Comment (XiaocongAi) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:647]: The comments need some update.

**Review by XiaocongAi — COMMENTED** [review]
Hi @paulgessinger , thank you for the clean-up and improvement. I have a few minor comments. Have you also looked at the CKF performance, e.g. efficiency, after those changes?

**Comment by paulgessinger** [comment]
Hey @XiaocongAi, thanks for looking at this. These specific changes here should give binary-identical results compared to the current main branch. I tested this locally, and the CI root file hash comparisons also succeed (and I didn't have to modify the hashes). There were some changes introduced by #1054, but those are separate and understood.

In terms of runtime performance, I ran it a number of times and looked at the timing, and I didn't observe any change in performance above noise.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp (hunk 2 above)
Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:647]: Should be updated.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:602]: Good idea. Factored out two functions `createSourceLinkTrackStates` and `processSelectedTrackStates`.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 3]
```diff
@@ -149,6 +200,10 @@ struct CombinatorialKalmanFilterResult {
   // Fitted states that the actor has handled.
   MultiTrajectory fittedStates;
 
+  // This is used internally to store candidate trackstates
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:203]: Removed.

File: Core/src/TrackFinding/MeasurementSelector.cpp [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    if (endIterator->chi2() >= chi2CutOff) {
+      break;  // endIterator now points at the first track state with chi2
+              // larger than our cutoff => defines the end of our returned
+              // range
+    }
+  }
+  ACTS_VERBOSE("Number of selected measurements: " << std::distance(
+                   candidates.begin(), endIterator));
+
+  // auto maxMeasIterator = std::next(candidates.begin() + nFinalCandidates);
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/src/TrackFinding/MeasurementSelector.cpp]: Totally right, should be gone.

File: Core/src/TrackFinding/MeasurementSelector.cpp [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+  // }
+  // candidates.push_back(trackStatesCopy.at(id));
+  // nRecorded++;
+  // }
+
+  isOutlier = false;
+  return std::pair{candidates.begin(), endIterator};
+}
+
+#else
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/src/TrackFinding/MeasurementSelector.cpp]: No we don't, I forgot to remove it. Gone now.

**Review by XiaocongAi — APPROVED** [review]
Hi @paulgessinger , thank you for the update and the clarification. I'm very happy with all the changes. Approved.

**Pull request merged**